pkg/schedule/affinity/group.go

```diff
@@ -0,0 +1,380 @@
+		return groupExpired
+	}
+	return groupDegraded
+}
+
+// SetState updates state and degradedExpiredAt with the following behavior:
+// - When the current state is groupAvailable
+//   - newState = groupAvailable: do nothing
+//   - newState = groupDegraded: change the state to groupDegraded and set degradedExpiredAt based on the current time
+//   - newState = groupExpired: change the state to groupExpired
+//
+// - When the current state is groupDegraded
+//   - newState = groupAvailable: change the state to groupAvailable
+//   - newState = groupDegraded: do nothing and do not refresh degradedExpiredAt
+//   - newState = groupExpired: change the state to groupExpired
+//
+// - When the current state is groupExpired
+//   - newState = groupAvailable: change the state to groupAvailable
+//   - newState = groupDegraded: do nothing and do not refresh degradedExpiredAt
+//   - newState = groupExpired: do nothing
+func (g *runtimeGroupInfo) SetState(newState condition) {
+	// If the expiration time has been reached, change groupDegraded to groupExpired.
+	if g.state == groupDegraded && g.IsExpired() {
+		g.state = groupExpired
+	}
+	// Update state
+	newState = newState.toGroupState()
+	if newState == groupDegraded {
+		// Only set the expiration time when transitioning from groupAvailable to groupDegraded.
+		// Do nothing if the original state is already groupDegraded or groupExpired.
+		if g.state == groupAvailable {
+			g.state = groupDegraded
+			g.degradedExpiredAt = newDegradedExpiredAtFromNow()
+		}
+	} else {
+		g.state = newState
+	}
+}
+
+func newDegradedExpiredAtFromNow() uint64 {
+	return uint64(time.Now().Unix()) + defaultDegradedExpirationSeconds
+}
+
+// IsAffinitySchedulingEnabled indicates whether affinity scheduling is allowed.
+func (g *runtimeGroupInfo) IsAffinitySchedulingEnabled() bool {
+	return g.IsAvailable() && g.LeaderStoreID != 0 && len(g.VoterStoreIDs) != 0
+}
+
+// IsRegularSchedulingEnabled indicates whether balance scheduling is allowed.
+func (g *runtimeGroupInfo) IsRegularSchedulingEnabled() bool {
+	return g.IsExpired() || g.LeaderStoreID == 0 || len(g.VoterStoreIDs) == 0
+}
+
+// AdjustGroup validates the group and sets default values.
+func (m *Manager) AdjustGroup(g *Group) error {
+	if err := ValidateGroupID(g.ID); err != nil {
+		return err
+	}
+	// If no distribution is provided, we will use the default distribution.
+	if g.LeaderStoreID == 0 && len(g.VoterStoreIDs) == 0 {
+		return nil
+	}
+	// Once provided, leader store ID and voter store IDs must be provided together.
+	if g.LeaderStoreID == 0 || len(g.VoterStoreIDs) == 0 {
+		return errs.ErrAffinityGroupContent.FastGenByArgs("leader store ID and voter store IDs must be provided together")
+	}
+
+	voterStoreIDs := slices.Clone(g.VoterStoreIDs)
+	slices.Sort(voterStoreIDs)
+	if slice.HasDupInSorted(voterStoreIDs) {
+		return errs.ErrAffinityGroupContent.FastGenByArgs("duplicate voter store ID")
+	}
+	if !slices.Contains(voterStoreIDs, g.LeaderStoreID) {
```

Comment:
In generative speaking, the leader should belong to the voters. How about renaming voterStoreIDs as followerStoreIDs if you don't allow the leader to exist in this list.